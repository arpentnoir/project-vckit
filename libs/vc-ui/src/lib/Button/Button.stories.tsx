import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { Button } from './Button';

const Story: ComponentMeta<typeof Button> = {
  component: Button,
  title: 'Button',
};
export default Story;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Contained = Template.bind({});
Contained.args = { label: 'Contained' };

export const Outlined = Template.bind({});
Outlined.args = { label: 'Outlined', variant: 'outlined' };
