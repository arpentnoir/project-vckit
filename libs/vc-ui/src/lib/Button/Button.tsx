import {
  Button as MUIButton,
  ButtonProps,
  CircularProgress,
  Stack,
  TypographyProps,
} from '@mui/material';
import React from 'react';
import { Text } from '../Text';

export interface IButton extends ButtonProps {
  label: string;
  textProps?: TypographyProps;
  loading?: boolean;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
}

export const Button = React.forwardRef<HTMLButtonElement, IButton>(
  (
    {
      label,
      textProps,
      variant = 'contained',
      color = 'primary',
      loading = false,
      leftIcon,
      rightIcon,
      ...rest
    },
    ref
  ) => {
    return (
      <MUIButton
        variant={variant}
        color={color}
        data-testid={`button:${label}`}
        {...rest}
        disabled={loading || rest.disabled}
        sx={{
          textTransform: 'none',
          padding: '8px 22px',
          width: 'fit-content',
          zIndex: 10,
          ...rest.sx,
        }}
        ref={ref}
      >
        {loading ? (
          <CircularProgress
            data-testid={`button:${label}:loader`}
            size="16px"
            sx={{ position: 'absolute' }}
            color={variant === 'contained' ? 'inherit' : color}
          />
        ) : null}
        {/* In order to maintain the size of the button when loader appears
      we set the label to be hidden
      */}
        <Stack
          visibility={loading ? 'hidden' : 'visible'}
          direction="row"
          justifyContent="center"
          alignItems="center"
          spacing="4px"
        >
          {leftIcon}
          <Text {...textProps} data-testid={`button:${label}:label`}>
            {label}
          </Text>
          {rightIcon}
        </Stack>
      </MUIButton>
    );
  }
);
