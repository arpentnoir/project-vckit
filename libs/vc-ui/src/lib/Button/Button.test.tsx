import '@testing-library/jest-dom';
import { getByTestId, queryByTestId, render } from '@testing-library/react';
import { Button } from './Button';

describe('Button', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<Button label="FooBar" />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should show spinner and hire label when loading', () => {
    const { baseElement, rerender } = render(<Button label="FooBar" />);

    expect(queryByTestId(baseElement, 'button:FooBar:loader')).toBeFalsy();
    expect(getByTestId(baseElement, 'button:FooBar:label')).toHaveStyle(
      'visibility: visible'
    );

    rerender(<Button label="FooBar" loading={true} />);

    expect(queryByTestId(baseElement, 'button:FooBar:loader')).toBeTruthy();
    expect(getByTestId(baseElement, 'button:FooBar:label')).toHaveStyle(
      'visibility: hidden'
    );
  });
});
