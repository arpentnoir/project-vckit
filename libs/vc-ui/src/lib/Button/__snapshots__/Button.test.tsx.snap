// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1yt9wte-MuiButtonBase-root-MuiButton-root"
      data-testid="button:FooBar"
      tabindex="0"
      type="button"
    >
      <div
        class="css-doi3ij-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
          data-testid="button:FooBar:label"
          tabindex="0"
        >
          FooBar
        </p>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;
